### PR TITLE
Feature Offline Sync groups

### PR DESCRIPTION
**Pull request by therajanmaurya**
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Apply the `MifosStyle.xml` style template to your code in Android Studio.

- [x] Run the unit tests with `./gradlew check` to make sure you didn't break anything

- [ ] If you have multiple commits please combine them into one commit by squashing them.

GroupList Enhance.
GroupDetails API Request Enhance.
Offline Sync Group Details and their accounts.
Database Encryption added.

### CODE REVIEW

File: mifosng-android/src/main/java/com/mifos/api/datamanager/DataManagerGroups.java [hunk 1]
```diff
@@ -52,14 +53,7 @@ public DataManagerGroups(BaseApiManager baseApiManager,
...
             case 0:
-                return mBaseApiManager.getGroupApi().getGroups(paged, offset, limit)
-                        .concatMap(new Func1<Page<Group>, Observable<? extends Page<Group>>>() {
-                            @Override
-                            public Observable<? extends Page<Group>> call(Page<Group> groupPage) {
-                                mDatabaseHelperGroups.saveGroups(groupPage);
-                                return Observable.just(groupPage);
-                            }
-                        });
+                return mBaseApiManager.getGroupApi().getGroups(paged, offset, limit);
```
Comment (ishan1604) [inline mifosng-android/src/main/java/com/mifos/api/datamanager/DataManagerGroups.java]: Good work pulling this code into a separate method.

File: mifosng-android/src/main/java/com/mifos/api/local/MifosSQLCipherHelper.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+public class MifosSQLCipherHelper extends SQLCipherOpenHelper {
+
+    public MifosSQLCipherHelper(DatabaseDefinition databaseDefinition,
+                                DatabaseHelperListener listener) {
+        super(databaseDefinition, listener);
+    }
+
+    @Override
+    protected String getCipherSecret() {
+        return "dbflow-rules";
```
Comment (ishan1604) [inline mifosng-android/src/main/java/com/mifos/api/local/MifosSQLCipherHelper.java]: I don't know how this CipherSecret works in DBFlow but don't you think we should generate it differently for each App Install.

Let's say we have a function that takes certain inputs and returns an output. For a given input the output should always be the same. 

Here is how it would look

SOME VALUE XYZ which is unique per device (Android user) ----> Function -----> Output is some hash

use this Hash as a Cipher Secret. 

So that it becomes even more difficult to decode the encrypted DB. 

Makes sense ?

File: mifosng-android/src/main/java/com/mifos/api/local/databasehelper/DatabaseHelperGroups.java [hunk 3]
```diff
@@ -34,50 +38,134 @@ public DatabaseHelperGroups() {
...
+                        .from(Group.class)
+                        .where(Group_Table.id.eq(groupId))
+                        .querySingle();
+
+                if (group != null) {
+                    group.setActivationDate(Arrays.asList(group.getGroupDate().getDay(),
+                            group.getGroupDate().getMonth(), group.getGroupDate().getYear()));
+                }
+
+                return null;
```
Comment (ishan1604) [inline mifosng-android/src/main/java/com/mifos/api/local/databasehelper/DatabaseHelperGroups.java]: Why are we returning null here ?

File: mifosng-android/src/main/java/com/mifos/api/local/MifosSQLCipherHelper.java (hunk 2 above)
Comment (therajanmaurya) [inline mifosng-android/src/main/java/com/mifos/api/local/MifosSQLCipherHelper.java]: Yeah I Agree. Can I do this in separate PR.

File: mifosng-android/src/main/java/com/mifos/api/local/databasehelper/DatabaseHelperGroups.java (hunk 3 above)
Comment (therajanmaurya) [inline mifosng-android/src/main/java/com/mifos/api/local/databasehelper/DatabaseHelperGroups.java]: Sorry for that When I have written this method, demo Server was down, So I have not test this method, Now fixed.

File: mifosng-android/src/main/java/com/mifos/api/local/MifosSQLCipherHelper.java (hunk 2 above)
Comment (ishan1604) [inline mifosng-android/src/main/java/com/mifos/api/local/MifosSQLCipherHelper.java]: Sure

Comment (ishan1604) [inline mifosng-android/src/main/java/com/mifos/api/local/MifosSQLCipherHelper.java]: If we are to add it in a separate PR, do all the Cipher work together because if it gets squashed in this one it will be difficult to track in future. so remove the cipher related code from this pr.

Comment (therajanmaurya) [inline mifosng-android/src/main/java/com/mifos/api/local/MifosSQLCipherHelper.java]: Ok, I agree. I will do that.

Comment (therajanmaurya) [inline mifosng-android/src/main/java/com/mifos/api/local/MifosSQLCipherHelper.java]: It's done @ishan1604. I have removed the SQLchiper code and now everything good. you can merge.

File: mifosng-android/src/main/java/com/mifos/api/local/databasehelper/DatabaseHelperGroups.java (hunk 3 above)
Comment (therajanmaurya) [inline mifosng-android/src/main/java/com/mifos/api/local/databasehelper/DatabaseHelperGroups.java]: It is fixed in below commit.
